docs/proposals/advanced-nginx-extensions.md

```diff
@@ -26,3 +74,915 @@ NGINX use cases. To allow them to implement those use cases, we need to bring a
+        group: example.org/v1alpha1
+        kind: RateLimitingFilter
+        name: rate-limit
+    backendRefs:
+    - name: headers
+      port: 80
+```
+
+Similarly, such filters can also be used in
+GRPCRoute. [TLSRoute](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.TLSRoute)
+doesn't support filters, so it needs to be added to the Gateway API first.
+
+TO-DO(pleshakov) - Add support for filters to the API section in the GEP after discussions about it.
+
+### Personas
+
+- Cluster operator creates a SnippetsTemplate.
+- Application developer uses it by creating a Custom resource of the values CRD of the SnippetsTemplate.
+
+### Validation
+
+#### NGINX Values
+
+When Cluster operator defines a values CRD, they should define validation rules to prevent invalid or malicious values
+propagating into NGINX config. The Kubernetes API server will perform that validation when an Application developer
+creates
+a Custom resource of the CRD.
+
+#### Template
+
+A template can be invalid (not follow go template or panic when executed). NGF must reject such templates
+and also not crash when executing them.
+
+A template can generate invalid NGINX configuration. When this happens, NGINX will continue to use the last valid
+configuration.
+However, any subsequent configuration updates (for example, caused by changes to an HTTPRoute) will not be possible
+until the invalid configuration is removed. Thus, Cluster operator must carefully design the template.
+
+### Security Considerations
+
+As mentioned in the previous section, Cluster operator should define validation rules to prevent invalid/malicious
+values from
+Custom resources created by Application developers from propagating into the NGINX config.
+
+However, it is possible that a template generates malicious values. Because of that, only Cluster operator (a privileged
+user) should be able to create SnippetsTemplates.
+
+### Upgrades
+
+NGF inserts configuration generated from the templates of a SnippetsTemplate into various parts of the configuration
+that it already generates. As a result, the configuration generated from the templates must play well with
+the surrounding configuration, so that it doesn't break that configuration and vice versa.
+
+As we develop NGF, that surrounding configuration will expand (because we will implement more NGINX features) and also
+might change (for example, due to improvements). As a result, there is a risk that SnippetsTemplate break after
+Cluster operator upgrades NGF to the next version. Such risk shall be clearly documented in the SnippetsTemplate
+documentation.
+
+### Prior Arts
+
+- NGINX Ingress
+  Controller [Custom annotations](https://docs.nginx.com/nginx-ingress-controller/configuration/ingress-resources/custom-annotations/)
+- NGINX Instance
+  Manager [Config Templates](https://docs.nginx.com/nginx-management-suite/nim/about/templates/config-templates/)
+
+### Alternatives
+
+- SnippetsPolicy
```

Comment:
Are you proposing that we choose either SnippetsPolicy or SnippetsTemplate? I feel like we should adopt both.

Reply:
both. 

I though it was clear from the intro:

```
## Advanced Extensions

This proposal brings two extension mechanisms:
. . .
```

Let me know how I can improve that

Reply:
I think it's just the alternatives sections that were confusing me. IMO alternatives should list the options that we decided not to pursue.